### DISCUSSION TITLE
Link to Google Colab button in Notebooks

### DISCUSSION THREAD

**Discussion by hollandjg** [AutoResearch/autora#501] [Show and Tell]
I wanted to highlight a feature of the documentation – the "Open in Google Colab" button.

The code builds on mkdocs-jupyter:
https://github.com/AutoResearch/autora/blob/e20b651a67adaa578e434ff0a7404545c621aee8/mkdocs.yml#L115

It is configured using the following lines (as of autora 3.1.0):
https://github.com/AutoResearch/autora/blob/e20b651a67adaa578e434ff0a7404545c621aee8/mkdocs.yml#L8-L9
This block specifies where the notebook can be found. 

> Note that we're using the "gh-pages" branch version of the notebook – this is the same branch where the documentation itself is stored, so it's guaranteed to be the same as the notebook version which the person is viewing in the docs.

The feature needs two "hooks" which convert strings, loaded here:
https://github.com/AutoResearch/autora/blob/e20b651a67adaa578e434ff0a7404545c621aee8/mkdocs.yml#L136-L137

... and implemented here: 
https://github.com/AutoResearch/autora/blob/e20b651a67adaa578e434ff0a7404545c621aee8/mkdocs/hooks.py#L5-L12

The actual link and its appearance are specified here:
https://github.com/AutoResearch/autora/blob/e20b651a67adaa578e434ff0a7404545c621aee8/mkdocs/overrides/main.html#L9-L12

This block of HTML uses the two hooks in the snippet 
```
{{ page.file.src_uri | basename | quote }}
```
- start with the `src_uri` of the `file` which generated the `page`
- `basename`: get the notebook's filename
- `quote`: convert any special characters (like spaces) into allowable values for a URL. 

The button shown is specified here:
https://github.com/AutoResearch/autora/blob/e20b651a67adaa578e434ff0a7404545c621aee8/mkdocs/overrides/main.html#L11

The "Download notebook" button is very similar, but doesn't need to use a special starting URL – it can use the "canonical_url":
https://github.com/AutoResearch/autora/blob/e20b651a67adaa578e434ff0a7404545c621aee8/mkdocs/overrides/main.html#L5-L8

Note that this requires the `site_url` to be set [properly](https://www.mkdocs.org/user-guide/configuration/#site_url):
https://github.com/AutoResearch/autora/blob/e20b651a67adaa578e434ff0a7404545c621aee8/mkdocs.yml#L5

(Note here: we noticed that we could have made the code for the Colab link a little simpler if we'd have left the `/autora` part off the end of the `site_url`, and we were tempted to do that. However, that misuse of the `site_url` configuration parameter could have caused more problems down the line – then the `canonical_url` would have been set wrong in the rest of the site, and any plugins which rely on it would also have broken.)

## Potential for development

- Turn it into a plugin for mkdocs which extends mkdocs-jupyter (or add it as a PR to mkdocs-jupyter)
- Add a cleaner hook which does the full conversion, rather than chaining multiple hooks in the html file, perhaps using f"strings". 
- Allow for multiple links, like to [binder](https://mybinder.org).
- Allow configurable buttons.

**Comment by chadcwilliams** [comment]
Closing as this is now implemented, well done!
